# 06.UseACorrectFormOfTheDoubleCheckedLockingIdiom

以下代码在SonarLint上会给出`DoubleCheckedLockingShouldNotBeUsed`的缺陷警告。

```Java
public class Singleton {
    private static Singleton instance;

    private Singleton() {
    }

    public static Singleton getInstance() {
        if (instance == null) {                  // 第一次读
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();  // 第二次读
                }
            }
        }

        return instance;                         // 第三次读
    }
}
```

缺陷原因如下：

`第一次读`和`第三次读`无法避免**指令重排序**，导致第三次读仍然有可能返回null值。为了避免重排序，保证`happen-before`顺序，我们可以使用`volatile`修饰Singleton字段。

修改如下：

```Java
public class Singleton {
    private static volatile Singleton instance;

    private Singleton() {
    }

    public static Singleton getInstance() {
        if (instance == null) {
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();
                }
            }
        }

        return instance;
    }
}
```

从1.7版本开始，JVM对`synchronied`关键字进行了优化，SonarLint的建议是使用

There are multiple ways to fix this. The simplest one is to simply not use double checked locking at all, and synchronize the whole method instead. With early versions of the JVM, synchronizing the whole method was generally advised against for performance reasons. But synchronized performance has improved a lot in newer JVMs, so this is now a preferred solution. If you prefer to avoid using synchronized altogether, you can use an inner static class to hold the reference instead. Inner static classes are guaranteed to load lazily.

Noncompliant Code Example

```Java
@NotThreadSafe
public class DoubleCheckedLocking {
    private static Resource resource;

    public static Resource getInstance() {
        if (resource == null) {
            synchronized (DoubleCheckedLocking.class) {
                if (resource == null)
                    resource = new Resource();
            }
        }
        return resource;
    }

    static class Resource {

    }
}
```


Compliant Solution

```Java
@ThreadSafe
public class SafeLazyInitialization {
    private static Resource resource;

    public synchronized static Resource getInstance() {
        if (resource == null)
            resource = new Resource();
        return resource;
    }

    static class Resource {
    }
} 
```

## 参考资料

* http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

